### ISSUE TITLE
Why does const-prop refuse to read from statics that contain pointers?

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#70356]
const-prop checks statics for whether they contain relocations, and if they do, it refuses to read from them. It is not clear why that happens.

~~As discussed in https://github.com/rust-lang/rust/pull/70241#discussion_r396066858, it turns out that const-prop will sometimes permit reading from mutable global allocations. That could be a soundness issue in case those allocations are actually mutated at run-time; const-prop would then be propagating non-constant data.~~

~~Currently it is not entirely clear what those mutable allocations are that const-prop is reading from, all we have is an ICE because an unexpected error ("const-prop does not support reading from mutable global allocation") arose during validation. It is also unclear (to me, at least) why const-prop is doing validation at all.~~

Cc @oli-obk @wesleywiser

**Comment by RalfJung** [comment]
I just realized the condition is `allocation.mutability == Mutability::Mut || allocation.relocations().len() > 0`, so maybe the allocation in question was not even mutable but had relocations in it. I am testing that now.

I do not understand the argument given in the code for why relocations are a problem:
```
        // If the static allocation is mutable or if it has relocations (it may be legal to mutate
        // the memory behind that in the future), then we can't const prop it.
```
If the memory behind that pointer is mutable, we will notice that when *that* allocation is being accessed. Why do we have to bail out so early? Is it because the hook used to only be called for statics, not all global allocations?

**Comment by RalfJung** [comment]
Also the mystery about why validation is even happening is answered here:

**Comment by RalfJung** [comment]
Also the mystery about why validation is even happening is answered here:

https://github.com/rust-lang/rust/blob/1add455ec6f81045e7651c6225902823f5d4fbfa/src/librustc_mir/transform/const_prop.rs#L643

I think I understand why, conservatively, this does validation... can this ever fail? Validation is quite picky about the kinds of errors in expects can happen, and with const-prop having a machine that bails out early on many operations, that could be a problem. The reason validation is so picky is to make sure we get good errors, which const-prop does not care about (the errors are not shown to the user), but it re-uses the same machinery.

**Comment by RalfJung** [comment]
Okay so the actual problem turned out to be relocations. There is no bug here, we can reject all reads from mutable globals (phew).

What remains open is why we check for the presence of relocations at all. @wesleywiser you added that comment that I mentioned above, would be great if you could explain it a bit more. :)

**Comment by wesleywiser** [comment]
@oli-obk [wanted it](https://github.com/rust-lang/rust/pull/64419#discussion_r327982230):

> for now also bail out if the allocation has relocations. If a static contains a *mut i32 it may be legal to mutate the memory behind that (if not now, maybe in the future).

But I'm not quite sure why...

**Comment by oli-obk** [comment]
Basically I want to avoid the following (assume it compiles with more boilerplate) from optimizing `foo` to `unsafe fn foo() -> u32 { 42 }`

```rust
static A: UnsafeCell<u32> = UnsafeCell::new(42);
static B: &'static UnsafeCell<u32> = &A;
unsafe fn foo() -> u32 {
	*B.get()
}
```

Since we can't ever know what users are up to with their code, the presence of relocations was the easiest way to avoid any kind of troubles there.

Furthermore, assuming we just forbid raw pointers from being derefed in const prop, we'd still have problems with the following getting const propped::

```rust
let x = Cell::new(42);
let y = &x;
```

I want to prevent const prop from creating any kind of mutable global memory that we may end up pointing to. While we can probably devise a smart scheme, right now it seems better to just forbid any relocations.

**Comment by RalfJung** [comment]
@oli-obk that sounds to me like an indirect way to do the check that we can now do much more reliably and directly:

https://github.com/rust-lang/rust/blob/cdb50c6f2507319f29104a25765bfb79ad53395c/src/librustc_mir/transform/const_prop.rs#L285-L287

This is done for all accesses to globals (things stored in `tcx`). So if a constant global points to a mutable one, the check will fire when the mutable global actually gets accessed.

**Comment by oli-obk** [comment]
that's true. Makes more sense to handle it there.

**Comment by oli-obk** [comment]
But how does that help with the second example I gave?

**Comment by RalfJung** [comment]
The second example does not involve any global though? So nothing we do in that hook will have any effect? I am confused.^^

**Comment by oli-obk** [comment]
The problem is that it would magically create a global if we used const prop here. Basically the same problems apply that we have with promotion.

**Comment by RalfJung** [comment]
But again, what does the `before_access_global` hook have to do with that? It won't prevent creation of that global, just access to it.

I also cannot imagine how const-prop could do *anything* in your example... `x` doesn't have a known value. And in a case like
```rust
let x = Cell::new(42);
let y = &x;
let z = x.get();
```
it would be entirely okay to replace `z` by `42`.

Of course with
```rust
let x = Cell::new(42);
let y = &x;
some_function(y);
let z = x.get();
```
it would not be okay, but then that is exactly the same as using mutable references -- and it has nothing to do with globals. So I fails to see the relevance of any of this for the discussion at hand.

**Comment by oli-obk** [comment]
I'm not worried about `z`. I'm worried about const propping into `y`, so `x` is eliminated entirely. 

```rust
let x = Cell::new(42);
let y = &x;
let z = y.get();
```

optimized to

```rust
static X: &'static Cell<i32> = &Cell::new(42);
let y = X;
let z = y.get();
```

I'm not sure how we're supposed to prevent this without banning relocations in propagated constants.

**Comment by RalfJung** [comment]
I don't see why we would even try to do that, that seems like a really odd transformation? And certainly, if any code is responsible for getting things like that right, it's the code that *constructs new constants*, not the code that *reads from existing constants*?

> I'm not sure how we're supposed to prevent this without banning relocations in propagated constants.

You still didn't explain how the `before_access_global` prevents anything like that. It prevents *reading from* constants that have relocation, but you keep talking about *constructing new* constants that have relocations.

My confusion remains: what does anything you say have to do with what the issue is about, namely [this check](https://github.com/rust-lang/rust/blob/cdb50c6f2507319f29104a25765bfb79ad53395c/src/librustc_mir/transform/const_prop.rs#L288)?

**Comment by oli-obk** [comment]
Ah, that's where the confusion comes from. Let's start with a code example again:

```rust
use std::sync::atomic::{AtomicUsize, Ordering};

fn main() {
    static A: AtomicUsize = AtomicUsize::new(42);
    static B: &'static AtomicUsize = &A;
    let x = B;
	A.store(0, Ordering::Relaxed);
    let y = x.load(Ordering::Relaxed);
}
```

If we const prop the `load` at some point (or construct your own lockfree datastructure that doesn't use any intrinsics), we have now changed behaviour because we never read the runtime value of `A` (0) but instead the initializer value (42).

**Comment by RalfJung** [comment]
> If we const prop the load at some point (or construct your own lockfree datastructure that doesn't use any intrinsics), we have now changed behaviour because we never read the runtime value of A (0) but instead the initializer value (42).

Indeed, that would be a problem. And that problem is caught by

**Comment by RalfJung** [comment]
(Sorry for the double-post, GH keeps submitting comments for me way before I am done. Probably some silly trigger-happy shortcut.)

> If we const prop the load at some point (or construct your own lockfree datastructure that doesn't use any intrinsics), we have now changed behaviour because we never read the runtime value of A (0) but instead the initializer value (42).

Indeed, that would be a problem. And that problem is caught by 

https://github.com/rust-lang/rust/blob/6c19a10e24af157b96687ca8dc1b48ebac4b9489/src/librustc_mir/transform/const_prop.rs#L285-L287

which would reject and read from a mutable global allocation such as `A`.

Which *additional* problem necessitates the additional check that follows next?

https://github.com/rust-lang/rust/blob/6c19a10e24af157b96687ca8dc1b48ebac4b9489/src/librustc_mir/transform/const_prop.rs#L288-L290

**Comment by oli-obk** [comment]
Oh. I see now. Yea, we don't need the second check, any later const prop will do the mutability check and bail out, no matter the number and kind of indirections.

We can remove the `if static_def_id.is_some() && allocation.relocations().len() > 0 { ` check

**Comment by RalfJung** [comment]
That's what I thought, thanks. :)
